### DISCUSSION TITLE
v0.9.1

### DISCUSSION THREAD

**Discussion by Louisvdw** [Louisvdw/dbus-serialbattery#73] [Announcements]
* Fixes to better handle driver exit for VenusOS V2.80
* Python 3 fixes for Sinowealth BMS
* Daly owners can invert their current flow indicator with INVERT_CURRENT_MEASUREMENT flag.
* Populates Consumed Amp Hours
* Show when bank is offline after 10 failed replies

<hr /><em>This discussion was created from the release <a href='https://github.com/Louisvdw/dbus-serialbattery/releases/tag/v0.9.1'>v0.9.1</a>.</em>

**Comment by nhybrow** [comment]
Hi,
How to use INVERT_CURRENT_MEASUREMENT flag ?
I put it in the utils.py file, with = -1 argument, but doesn't work
thanks in advance

**Comment by Louisvdw** [comment]
Just to confirm you are using a normal Daly BMS. Not a Sinowealth chip version?

**Reply by nhybrow** [reply]
I dont find Sinowealth anywhere in the venus information of my DalyBMS. I think so it's not a Sinowealth.
https://photos.app.goo.gl/N8p9DRx7CSG8pbQN8

**Reply by Louisvdw** [reply]
That is the normal Daly, so it should work. Let me have a look

**Comment by mendok** [comment]
Hello and merry Xmas. 
Just wondering if and when a new version to include cerbo gx fix??
As i am having basic problems logging into my cerbo gx.
Keeps saying access denied 
Cheers mendo

**Reply by Louisvdw** [reply]
The problem with the Cerbo GX where it does not run the autoinstaller is an issue with the device and the VenusOS software. This will be something the Victron team will be looking into. The driver just use that Victron service as an install option. Until this is fixed in a new firmware you will have to use one of the other install methods.

**Comment by Martin1683** [comment]
Hallo Louis,

wundervolles Projekt vielen lieben Dank dafür. Eine Spende ist unterwegs. Eine Frage hätte ich aber: Immer wieder lese ich, dass es manchmal Probleme gibt, wenn zwei USB-Treiber aktiv sind, z.B. ein Gridmeter von Carlo Gavazzi

Ich habe einen Cerbo GX. Wie kann ich den Treiber eigentlich wieder entfernen, falls es Probleme geben sollte? Oder ist der Treiber nur aktiv, wenn auch ein USB - RS232/485 Kabel angeschlossen ist? Hilft es bei Probleme auf Werkseinstellung zurückzusetzen?

Vielen Dank für deine Info.

MFG Martin

**Reply by Louisvdw** [reply]
Hi Martin
The driver only runs on serial ports, so you have have a serial port active before it will start up. The driver get started very last after all others could not connect, so the only reports I had with a Carlo Gavazzi meter acting up is when that meter lost utility power. This olso only happens on the V2.80 VenusOS firmware, so you will be fine on V2.73.
Lastly to uninstall is very easy. There is a script you can run to disable the driver for you. `sh /data/etcdbus-serialbattery/disabledriver.sh`

**Reply by Martin1683** [reply]
Hallo Louis,

vielen Dank für die Rückmeldung. Kannst du mir noch sagen, mit welchem Skript man den Treiber wieder aktivieren kann, wenn er deaktiviert wurde. Wird er automatisch aktiviert, wenn man ein Update macht?

Noch eine Frage, da ich leider kein Programmierer bin und ich mit Venus keine Erfahrung habe:

Wenn ich z.B. die MAX_BATTERY_DISCHARGE_CURRENT oder eine einen anderen Parameter ändern möchte, wie mache ich das? Wird das in Putty geändert oder wird die Datei im Editor unter Windows verändert?

Gibt es Infoseite im Web, wo man die Programmierbefehle lernen kann?

MFG Martin

**Reply by Louisvdw** [reply]
Hi Martin
Once you have disabled the driver you can just install it again to enable it. 
To change the settings you can edit the utils.py file while using SSH using the nano tool, or you can edit that file on your PC and copy it over to the GX device.
I'll update the wiki to explain it in more detail.

**Reply by Martin1683** [reply]
Hallo Louis,

was ich auch nicht verstehe sind die angezeigten Werte, siehe Bild

![Parameter](https://user-images.githubusercontent.com/97060744/148355921-89b53e10-c1af-471c-9fba-bbbe6b1fb0c8.JPG)
.
Die Werte kommen nicht von der utils.py und auch nicht vom BMS. Woher kommen die Werte?

**Reply by Louisvdw** [reply]
Hi Martin

Those are described in the [Wiki under Features](https://github.com/Louisvdw/dbus-serialbattery/wiki/Features) and under How to Install.

```
battery parameters
Charge Voltage Limit(CVL)
Charge Current Limit(CCL)
Discharge Current Limit(DCL)
CVL (Battery Max) automatically adjusted by cell count * 3.45V
Battery Min automatically adjusted by cell count * 3.1V
```
```
Edit the source to update the constants. /data/etc/dbus-serialbattery/utils.py
MAX_BATTERY_CURRENT = 50.0
MAX_BATTERY_DISCHARGE_CURRENT = 60.0
```

So the charge and discharge current limits come from the settings in utils.py while the charge voltage is calculated by by the amount of cells that your battery consist of. For your example of 55.2V that would be for 16 cells.
If you want to adjust this then you can edit utis.py
```
MAX_CELL_VOLTAGE = 3.45
```

**Reply by Martin1683** [reply]
Hallo Louis,

ja jetzt bin ich auf dem Schlauch gestanden. Bei "between 30% - 35% => 1/2 Max discharge" --> 60.0 /2 = 30.0A ;-)

Ich werde mit den Werten noch warten bis Sie die Wiki ergänzt haben, da ich nicht weiß, wie ich die Datei ändere bzw. auf das Venus bekomme.

Vielen Dank für alles. Grüße Martin

**Reply by Martin1683** [reply]
Hallo Louis,

jetzt hab ich mich doch ein bißchen in Linux eingelesen. Hat alles mit dem Tool "Nano" geklappt. War auch nicht schwer!

MFG Martin

**Reply by Martin1683** [reply]
Hallo Louis,

leider brauche ich nochmals ihre Hilfe. 

Zuerst mal zu meinem System:
Lifepo4 200Ah 16 Stück mit Daly BMS 16S 150 A
Victron Multiplus ll 48/5000/70-50
Cerbo GX
Mppt 150/85
ESS-System enabled

Ist ihr CCCM universell anwendbar bzw. Kapazitätsunabhängig und für alle Lifepo4-Zellen geeignet?

Was mir beim Daly BMS aufgefallen ist: Ströme unter ca. 2A werden nicht erkannt. Wenn also in der Nacht ca. 100 W verbraucht werden (100W/52V =ca. 1,92A) Ist das ein Problem, wenn der Stromfluss nicht erkannt wird? Der Multiplus erkennt das besser, ist aber wahrscheinlich zur SOC-Berechnung nicht gut geignet oder?

Das ESS berücksichtigt ja auch immer die SOC tiefe. Im Manual steht, dass die max. 80 % Kapazität genutzt werden. Sollen die fixen Spannungen in der Utily.py dann ausgereizt werden z.B. max = 3,55V und min = 2,7V. Bringt das was, oder wird der SOC im Daly ganz anders berechnet?

Bestimmt der Treiber eigentlich auch die Ladekurve? Ich meine damit Konstantspannungsphase und Konstantstromphase? Oder kommen die aus dem Multiplus? Siehe z.B. Bild im Anhang (Das sind nicht meine Einstellungen, sind aus dem Internet)

Vielen Dank für deine Antworten.

Grüße Martin



![VE Configure LiFeYPO-600AH-12V -4](https://user-images.githubusercontent.com/97060744/148607214-b2625f96-a037-4817-ba04-9f0da7b4e2e6.jpg)

**Reply by Louisvdw** [reply]
Hi Martin

The CCCM works with your battery's values. So it will 1/2 or 1/4 the values looking at the SOC. But this is also only on the edges of where the battery is used and it helps that a sudden pull/push of power does not push the lithium cells over their capacity and jumping in their voltage. It's safe for any lithium cell battery. In between the edges you will always have full capability (current).

Working with small currents are problematic for most BMS systems. The shunts measuring these current have to be calibrated for either high or low currents, which means if you want a BMS that can acurately measure both high and low current you will need to have 2 shunts  in the BMS. Most don't and most users also choose cheaper options which does not have that. The Daly does not have the most acurate low current shunt. The JBD I have struggle with anything under 0.8A as well. 
But the BMS are still the more acurate SOC meaurement. Infact the value you get from your Multiplus II is actually calcualted from the AC side, so it is a estimate on the DC side.
The BMS does have some limits on cell voltages that it will use to correct itself if the SOC drift to far off. It does this on a per cell measurement whitch is better than any other device can do that does not measure on a cell level.

SOC calculations are different based on what inputs and formulas the BMS has. The driver only translate that into the GX device. In many BMS you can edit some of these parameters. In the Daly you cannot.

A charging curve is something that lead acid batteries use, and as lead acid batteries do not have a management system any charging device (like a multiplus or MPPT) will have to have a charging curve. There is very little voltage drop on litium cells over their charging state, while on lead acids you can measure the cells charge state from that voltage. 
So for litiums you need a BMS that can manage this process and this also include the charging. 
So for lithiums you would normally set the Absorption and Float values for charging the same of very close to each other. The BMS will then manage the charging of each cell and the balancers will work to keep each cell's value the same. 
In the driver we set the max cell voltage and this determine the voltage we aim for. Once that voltage is reached the battery will be kept at that level and this will be your full mark (SOC 100%).

**Reply by Martin1683** [reply]
Hallo Louis,

vielen Dank für die Antworten. Jetzt bin ich definitiv schlauer. Vielen Dank nochmal für alles und mach weiter so.

Grüße aus Bayern.

Martin

**Reply by Martin1683** [reply]
Hallo Louis,

da sie ein Profi sind und ich Ihnen sehr vertraue, möchte ich Sie noch fragen, was Sie von aktiven Balancern halten? Ich habe mir so etwas gekauft, siehe Anhang
[5A-active-balancer-manual-NEW-0928.pdf](https://github.com/Louisvdw/dbus-serialbattery/files/7833851/5A-active-balancer-manual-NEW-0928.pdf)

Ist es sinnvoll, so etwas einzusetzten? Leider sagt im Internet jeder etwas anderes.

MFG Martin
.

**Reply by Martin1683** [reply]
Hallo Louis,

ich überlege, deinen Treiber im deutschsprachigen Raum populär zu machen und ein Youtube-Video zu machen, da ich absolut genial finde, was du programmiert hast. So ein System kostet dem Mainstream viel Geld. Wäre das für dich Ok oder möchtest du das eher nicht?

Grüße Martin

**Reply by Louisvdw** [reply]
Hi Martin

You are welcome to share information about the driver. I love to hear from people that find it useful and see their system that they use it in.

About balancers. Balancers are great to have for lithium batteries. They make sure that one cell does not get to far out of sync with the rest of the battery. And you will find that all BMS have them. It is part of what makes your battery last much longer by keeping your cells where they should be.

The size of the balancers and active vs. passive balancers is where peoples preverences differ. With that it is all about what you need and how much you are willing to spend.

In theory when your battery's cells are all matched up and balanced they should stay that way.
However no 2 cells are exactly the same. They have some manufacturing tolerances that make them all unique. It is these differences that make one cell deplete faster or another charged faster. 
If you get matched cells for your battery, meaning they have very close to the same charactiristics, and you start with them on all the same charge levels then you only need small balancers to keep them the same. 
If you start using your battery before the cells have the same charge, then your balancers have to work so much harder to get all the cells the same and this can take a  week or two. Then larger balancers can do this faster. But after the cells are all the same you again only need very small balancers and a large one is a waste.

Active balancers will move the energy from the high cells to the low cells, while pasive balancers will use a load to reduce the voltage on e the high cells until the low cells can catch up. 
Passive use less complex design which is cheaper to make, while active will save you energy, but cost more.

Would I add active balancers on top of my BMS? Propably not.
But if I do have active balancers I would add it to the battery when it is build to reduce the time balance them and remove them afterwards.

I have seen quite a few instances when the BMS did not correctly keep the battery balanced. In such a case I can see that an external balancer will be better. But this is a case of fixing a problem for spesific hardware and would not be a generic advice.
The Daly BMS and Tian Power have issues with some of their balance logic in some models. It is less of an issue if the BMS is connected to the driver as the charge system and the BMS can now work together.

**Reply by Martin1683** [reply]
Hallo Louis,

was mir aufgefallen ist: Das CCCM sollte eigentlich bei SOC >98% auf 1A Ladestrom begrenzen. Im Rahmen des ESS wenn der Multiplus die Batterien auflädt, wird nichts begrenzt. Wie kann das sein
![Unbenannt](https://user-images.githubusercontent.com/97060744/148814714-502cd056-4b99-41cc-a703-440b114e5323.JPG)
.

**Reply by Louisvdw** [reply]
The CCCM formulas had been updated in a earlier release to allow more (5A in the 99-100% range), but the documentation was outdated. I fixed the wiki with the new CCCM limits.
The other problem you will see is that the multi will struggle keeping the limits exact in the lower current. That is just cause it is less acurate with low current than with high current. But is should close the correct range.

**Reply by Martin1683** [reply]
Hallo Louis,

jetzt muss ich dich noch was fragen, da ich mir das nicht erklären kann. Im Multiplus habe 1 Stunde Absorptionszeit eingestellt. Der MPPT folgt ihm gemäß ESS Handbuch. Nur bleibt es nicht bei einer Stunde Absortionsspannung. Er geht nicht in Float über.

Wird das über deinen Treiber gesteuert?

MFG Martin

**Reply by Louisvdw** [reply]
For the driver you activate DVCC which means the battery now is in control, so your float and absorbtion settings will be ignored. 
It is a good thing as those are used for lead acid batteries so not needed anymore.
You can go read the DVCC section on the Victron website. It has a much more detailed explanation
